core/services/relay/evm/chain_reader.go

```diff
@@ -0,0 +1,202 @@
+package evm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+type ChainReaderService interface {
+	services.ServiceCtx
+	commontypes.ChainReader
+}
+
+type chainReader struct {
+	lggr       logger.Logger
+	contractID common.Address
+	lp         logpoller.LogPoller
+}
+
+// NewChainReaderService constructor for ChainReader
+func NewChainReaderService(lggr logger.Logger, lp logpoller.LogPoller, ropts *types.RelayOpts) (*chainReader, error) {
+	relayConfig, err := ropts.RelayConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing RelayConfig: %w", err)
+	}
+
+	if !common.IsHexAddress(ropts.ContractID) {
+		return nil, fmt.Errorf("%w: invalid contractID, expected hex address", commontypes.ErrInvalidConfig)
+	}
+	contractID := common.HexToAddress(ropts.ContractID)
+
+	if relayConfig.ChainReader == nil {
+		return nil, fmt.Errorf("%w: ChainReader missing from RelayConfig", errors.ErrUnsupported)
+	}
+
+	if err = validateChainReaderConfig(*relayConfig.ChainReader); err != nil {
+		return nil, err
+	}
+
+	return &chainReader{lggr.Named("ChainReader"), contractID, lp}, nil
```

Comment:
I see that we are not passing `relayConfig.ChainReader` yet. What is the expected functionality at this stage?

Reply:
Yeah it should be out of scope for this PR. I did a full implementation based off of this PR [here](https://github.com/smartcontractkit/chainlink/pull/11272), although its outdated now but will clean it up and make it use Ryans Codec after this is merged.